[8.0][ADD][base_import_match] Import deduplicator.

yajo opened the pull request:
By default, when importing data (like CSV import) with the `base_import`
module, Odoo follows this rule:
1. If you import the XMLID of a record, make an **update**.
2. If you do not, **create** a new record.

This module allows you to set additional rules to match if a given import is an
update or a new record.

This is useful when you need to sync heterogeneous databases, and the field you
use to match records in those databases with Odoo's is not the XMLID but the
name, VAT, email, etc.

After installing this module, the import logic will be changed to:
1. If you import the XMLID of a record, make an **update**.
2. If you do not:
   1. If there are import match rules for the model you are importing:
      1. Discard the rules that require fields you are not importing.
      2. Traverse the remaining rules one by one in order to find a match in
         the database.
         1. If one match is found:
            1. Stop traversing the rest of valid rules.
            2. **Update** that record.
         2. If zero or multiple matches are found:
            1. Continue with the next rule.
         3. If all rules are exhausted and no single match is found:
            1. **Create** a new record.
   2. If there are no match rules for your model:
      1. **Create** a new record.

Commit:
[8.0][ADD][base_import_match] Import deduplicator.

By default, when importing data (like CSV import) with the ``base_import``
module, Odoo follows this rule:

1. If you import the XMLID of a record, make an **update**.
2. If you do not, **create** a new record.

This module allows you to set additional rules to match if a given import is an
update or a new record.

This is useful when you need to sync heterogeneous databases, and the field you
use to match records in those databases with Odoo's is not the XMLID but the
name, VAT, email, etc.

After installing this module, the import logic will be changed to:

1. If you import the XMLID of a record, make an **update**.
2. If you do not:
   1. If there are import match rules for the model you are importing:
       1. Discard the rules that require fields you are not importing.
       2. Traverse the remaining rules one by one in order to find a match in
          the database.
          1. If one match is found:
             1. Stop traversing the rest of valid rules.
             2. **Update** that record.
          2. If zero or multiple matches are found:
             1. Continue with the next rule.
          3. If all rules are exhausted and no single match is found:
             1. **Create** a new record.
   2. If there are no match rules for your model:
      1. **Create** a new record.

dreispt commented on base_import_match/models/base_import.py:
```diff
@@ -0,0 +1,220 @@
+
+
+class BaseImportMatch(models.Model):
+    _name = "base_import.match"
```
Not that it makes a real difference, but shouldn't this be `base.import.match`?

yajo replied:
Core `base_import` module uses the `base_import` namespace for its models.

dreispt replied:
ok

dreispt commented on base_import_match/models/base_import.py:
```diff
@@ -0,0 +1,220 @@
+
+        return result
+
+    # TODO convert to @api.model_cr in v10
```
Why not do it now?

yajo replied:
`@api.model_cr` does not exist until v10.

dreispt replied:
ok

dreispt commented on base_import_match/models/base_import.py:
```diff
@@ -0,0 +1,220 @@
+                #    the corresponding exception.
+                # In any case, we must do nothing.
+                except ValueError:
+                    continue
```
Maybe this could be an option in the match rule: a checkbox for "Error if multiple matches found?"

yajo replied:
Good idea, but I'll add it to roadmap, it's not my current target.

dreispt replied:
Roadmap idea: add a log to write these exceptions to, so that end users can monitor it and decide what to do to each individual record: ignore or force load.

yajo replied:
That's too much IMO. I did not want to change the import UI with this, just the inner behavior.

dreispt commented on base_import_match/models/base_import.py:
```diff
@@ -0,0 +1,220 @@
+                    continue
+
+                # Give a valid XMLID to this row
+                row["id"] = match._BaseModel__export_xml_id()
```
Missing a dot: `match._BaseModel.__export_xml_id()`

dreispt replied:
Shouldn't it be just `match.__export_xml_id()` ?

yajo replied:
This must be what it is :wink:

dreispt commented:
I know this pain, and currently solve it by providing my own XML Ids at import time, generated based on the key attributes, such as the Customer VAT or a Contract code.

This is an Interesting alternative solution for that problem.
I felt the code somewhat hard to read at some points, and wish it could be clearer, but that doesn't make it wrong.

It's +1 for me once comments fixed and Travis is green.

Commit:
[IMP] Add conditionals to fields when importing.

This patch allows to import res.partner records by VAT when is_company==True by default.

yajo commented:
I know it's hard to read, but the import feature in Odoo is quite cumbersome. I had to dig a lot in the code to make this work.

I refactored it in c0f7ceb a bit to make it compatible with company import by VAT, and added that profile by default. Please test the new version, it's quite better.

Commit:
Add security for new models.

moylop260 commented:
Could you check rst-syntax?

``` bash
base_import_match/__init__.py:1: [E7901(rst-syntax-error), ] README.rst:30 Unexpected indentation.
base_import_match/__init__.py:1: [E7901(rst-syntax-error), ] README.rst:63 Unexpected indentation.
base_import_match/__init__.py:1: [E7901(rst-syntax-error), ] README.rst:59 Inline strong start-string without end-string.
```

More info here: https://travis-ci.org/OCA/server-tools/jobs/129966566#L430

yajo committed:
Try to fix linter errors.

dreispt commented:
:+1:

moylop260 commented:
@Yajo 
FYI I used http://rst.ninjs.org/ to fix [current errors](https://travis-ci.org/OCA/server-tools/jobs/132766029#L423-L427) of `base_import_match/README.rst` file.

And applying the following patch:

``` patch
diff --git base_import_match/README.rst base_import_match/README.rst
index f1c6577..52609b4 100644
--- base_import_match/README.rst
+++ base_import_match/README.rst
@@ -23,20 +23,26 @@ After installing this module, the import logic will be changed to:

 - If you import the XMLID of a record, make an **update**.
 - If you do not:
+
   - If there are import match rules for the model you are importing:
-      - Discard the rules that require fields you are not importing.
-      - Traverse the remaining rules one by one in order to find a match in
-        the database.
-          - Skip the rule if it requires a special condition that is not
-            satisfied.
-          - If one match is found:
-            - Stop traversing the rest of valid rules.
-            - **Update** that record.
-          - If zero or multiple matches are found:
-            - Continue with the next rule.
-          - If all rules are exhausted and no single match is found:
-            - **Create** a new record.
+
+    - Discard the rules that require fields you are not importing.
+    - Traverse the remaining rules one by one in order to find a match in the database.
+
+        - Skip the rule if it requires a special condition that is not
+          satisfied.
+        - If one match is found:
+
+          - Stop traversing the rest of valid rules.
+          - **Update** that record.
+        - If zero or multiple matches are found:
+
+          - Continue with the next rule.
+        - If all rules are exhausted and no single match is found:
+
+          - **Create** a new record.
   - If there are no match rules for your model:
+
     - **Create** a new record.

 By default 2 rules are installed for production instances:
@@ -59,6 +65,7 @@ To configure this module, you need to:
 #. If the rule must be used only for certain imported values, check
    *Conditional* and enter the **exact string** that is going to be imported
    in *Imported value*.
+
    #. Keep in mind that the match here is evaluated as a case sensitive
       **text string** always. If you enter e.g. ``True``, it will match that
       string, but will not match ``1`` or ``true``.

```

Now don't show rst errors:
<img width="1248" alt="screen shot 2016-05-25 at 10 40 00 am" src="https://cloud.githubusercontent.com/assets/6644187/15546318/0f175bd2-2265-11e6-9726-a114f64e8992.png">

moylop260 commented:
👍

dreispt commented:
:+1:

yajo commented:
@moylop260 please pr me with that

moylop260 commented:
You can use this patch with `git apply moy.patch` or `patch -p0 <moy.patch`
Just save it in a file.
Could you try it or you prefer a pr?

moylop260 committed:
[REF] base_import_match: Small rst fixes in readme file

moylop260 commented:
@Yajo [yajo#1](https://github.com/Yajo/server-tools/pull/1/files)

yajo committed:
Merge pull request #1 from vauxoo-dev/yajo-rst-fix-moy

[REF] base_import_match: Small rst fixes in readme file

yajo commented:
:green_heart:

dreispt commented on base_import_match/models/base_import.py:
```diff
@@ -0,0 +1,297 @@
+    def create(self, vals):
+        """Wrap the model after creation."""
+        result = super(BaseImportMatch, self).create(vals)
+        self._load_autopatch(result.model_name)
```
@Yajo I'm curious because I don't understand how this works:
AFAICS you are here specifically patching the BaseImportMatch model, not the BaseModel.
So this patch won't affect the other models as expected.
But probably I'm missing something.

yajo replied:
This only patches the modules found in configuration. That's why that method is called when creating, deleting, writing and restarting Odoo: it patches configured models and unpatches those that not.

dreispt replied:
OK, I can see.
